### ISSUE TITLE
New tabs using cmd-T in a git repository don't open with same directory in macOS Terminal

### ISSUE THREAD

**Issue by jidicula** [ohmyzsh/ohmyzsh#8263]
**Describe the bug**
When the current directory is also a git repository, new tabs created using cmd-T open in the home directory, instead of the current working directory, regardless of Terminal.app preferences. New tabs created from directories that are not git repositories open in the current working directory, as expected. This behaviour persists when all plugins are disabled.

When the `osx` plugin is enabled, the `tab` command in a git repository opens a new tab in the current working directory, as expected, but new tabs created using cmd-T still open in the home directory unexpectedly. Relying on the `tab` command isn't a good solution for when the user is in another interpreter (in my case Python or R) in a git repository and wants to open a new tab in the same directory.

**To Reproduce**

1. Ensure that Terminal preferences are set to open a new tab in the same working directory: Terminal -> Preferences -> General -> New tabs open with:
2. Open a new Terminal window.
3. `cd` to any git repository.
4. Open a new tab using cmd-T or New Tab from the Shell menu.
5. New tab will open in the home directory.

**Expected behavior**
The Terminal.app preference for new tabs opening in the same working directory should apply to any directory, not just non-git ones.

**Screenshots or recordings**

**System:**
 - OS: macOS 10.15
 - Zsh version 5.7.1
 - Terminal emulator: Terminal.app (macOS built-in)

**Additional context**
Possibly related to Issue #2646

**Comment by mcornella** [comment]
Can you check if the `update_terminalapp_cwd` function is getting called in git repositories? Run `typeset -f -t update_terminalapp_cwd` to enable tracing of it (this will print a bunch of output if the function ran).

**Comment by jidicula** [comment]
Sure, should I run this before or after opening a new tab?

**Comment by mcornella** [comment]
You should run it *before entering a git repository*, then enter such git repository. If there is output (post it), it means it ran, and opening a new tab should in theory open in the same directory.

**Comment by jidicula** [comment]
Here's the output (I modified the hostname and pathnames for privacy):
```zsh
johanan@hostname:~$ typeset -f -t update_terminalapp_cwd
+update_terminalapp_cwd:1> emulate -L zsh                                       
+update_terminalapp_cwd:4> omz_urlencode -P /Users/johanan
+omz_urlencode:1> emulate -L zsh
+omz_urlencode:2> zparseopts -D -E -a opts r m P
+omz_urlencode:4> local in_str=/Users/johanan
+omz_urlencode:5> local url_str=''
+omz_urlencode:6> local spaces_as_plus
+omz_urlencode:7> [[ -z -P ]]
+omz_urlencode:8> local str=/Users/johanan
+omz_urlencode:11> local encoding=UTF-8
+omz_urlencode:12> local safe_encodings
+omz_urlencode:13> safe_encodings=( UTF-8 utf8 US-ASCII ) 
+omz_urlencode:14> [[ -z UTF-8 ]]
+omz_urlencode:23> local i byte ord LC_ALL=C
+omz_urlencode:24> export LC_ALL
+omz_urlencode:25> local reserved=';/?:@&=+$,'
+omz_urlencode:26> local mark='_.!~*()-'
+omz_urlencode:27> local dont_escape='[A-Za-z0-9'
+omz_urlencode:28> [[ -z '' ]]
+omz_urlencode:29> dont_escape+=';/?:@&=+$,' 
+omz_urlencode:32> [[ -z '' ]]
+omz_urlencode:33> dont_escape+='_.!~*()-' 
+omz_urlencode:35> dont_escape+=']' 
+omz_urlencode:39> local url_str=''
+omz_urlencode:40> i = 1
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=/ 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=/ 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=U 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=U 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=s 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=s 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=e 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=e 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=r 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=r 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=s 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=s 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=/ 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=/ 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=j 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=j 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=o 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=o 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=h 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=h 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=a 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=a 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=n 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=n 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=a 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=a 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:41> byte=n 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=n 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 14
+omz_urlencode:53> echo -E /Users/johanan
+update_terminalapp_cwd:4> local URL_PATH=/Users/johanan
+update_terminalapp_cwd:5> [[ 0 != 0 ]]
+update_terminalapp_cwd:8> printf '\e]7;%s\a' file://hostname/Users/johanan
johanan@hostname:~$ cd path-to/git-repo 
+update_terminalapp_cwd:1> emulate -L zsh                                       
+update_terminalapp_cwd:4> omz_urlencode -P /Users/johanan/path-to/git-repo
+omz_urlencode:1> emulate -L zsh
+omz_urlencode:2> zparseopts -D -E -a opts r m P
+omz_urlencode:4> local in_str=/Users/johanan/path-to/git-repo
+omz_urlencode:5> local url_str=''
+omz_urlencode:6> local spaces_as_plus
+omz_urlencode:7> [[ -z -P ]]
+omz_urlencode:8> local str=/Users/johanan/path-to/git-repo
+omz_urlencode:11> local encoding=UTF-8
+omz_urlencode:12> local safe_encodings
+omz_urlencode:13> safe_encodings=( UTF-8 utf8 US-ASCII ) 
+omz_urlencode:14> [[ -z UTF-8 ]]
+omz_urlencode:23> local i byte ord LC_ALL=C
+omz_urlencode:24> export LC_ALL
+omz_urlencode:25> local reserved=';/?:@&=+$,'
+omz_urlencode:26> local mark='_.!~*()-'
+omz_urlencode:27> local dont_escape='[A-Za-z0-9'
+omz_urlencode:28> [[ -z '' ]]
+omz_urlencode:29> dont_escape+=';/?:@&=+$,' 
+omz_urlencode:32> [[ -z '' ]]
+omz_urlencode:33> dont_escape+='_.!~*()-' 
+omz_urlencode:35> dont_escape+=']' 
+omz_urlencode:39> local url_str=''
+omz_urlencode:40> i = 1
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=/ 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=/ 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=U 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=U 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=s 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=s 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=e 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=e 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=r 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=r 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=s 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=s 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=/ 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=/ 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=j 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=j 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=o 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=o 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=h 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=h 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=a 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=a 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=n 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=n 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=a 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=a 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=n 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=n 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=/ 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=/ 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=p 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=p 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=e 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=e 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=r 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=r 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=s 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=s 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=o 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=o 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=n 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=n 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=a 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=a 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=l 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=l 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=- 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=- 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=p 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=p 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=r 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=r 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=o 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=o 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=j 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=j 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=e 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=e 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=c 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=c 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=t 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=t 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=s 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=s 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=/ 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=/ 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=t 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=t 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=i 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=i 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=n 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=n 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=d 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=d 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=e 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=e 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=r 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=r 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=i 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=i 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=z 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=z 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=e 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=e 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:41> byte=r 
+omz_urlencode:42> [[ "$byte" -regex-match "$dont_escape" ]]
+omz_urlencode:43> url_str+=r 
+omz_urlencode:40> ++i 
+omz_urlencode:40> i <= 43
+omz_urlencode:53> echo -E /Users/johanan/path-to/git-repo
+update_terminalapp_cwd:4> local URL_PATH=/Users/johanan/path-to/git-repo
+update_terminalapp_cwd:5> [[ 0 != 0 ]]
+update_terminalapp_cwd:8> printf '\e]7;%s\a' file://hostname/Users/johanan/path-to/git-repo
```

I then tried to open a new tab using cmd-T and it did not open in the same directory.

**Comment by mcornella** [comment]
The relevant parts are (edited `omz_urlencode` function execution for clarity):
```
+update_terminalapp_cwd:8> printf '\e]7;%s\a' file://hostname/Users/johanan
+update_terminalapp_cwd:8> printf '\e]7;%s\a' file://hostname/Users/johanan/path-to/git-repo
```
I assume the git repositories don't have any special characters that need to be URL-encoded.

In both directories the printf with the non-documented sequence are executed, but it's only failing on directories that have a git repository in it... That's a weird bug, Terminal.app shouldn't care what it has inside. Did this start to happen after upgrading to Catalina?

You can make your own testing lab to pinpoint the exact cause of the bug. For starters, empty out the `update_terminal_app` function so that it doesn't run anymore, then create one you can use to set the new-tab directory manually without having to cd to it.
```
update_terminal_app(){}
# You'd use this like `set_newtabdir path/to/dir`
# Don't use it with a directory with spaces or other special characters
set_newtabdir() { printf '\e]7;%s\a' file://hostname/${1:a}; }
```
Then, you can make a bunch of directories and run the printf command with the appropriate file URI, and see if a new tab opens it there. One of the things to test is if it looks if it contains a .git directory, you don't have to `git init` it. Another thing you have to check is if using Oh My Zsh triggers it, so comment out the `source $ZSH/oh-my-zsh.sh` line from your zshrc and run the `set_newtabdir` function with the path to a git directory to test.

Let me know what you find.

**Comment by jidicula** [comment]
> I assume the git repositories don't have any special characters that need to be URL-encoded.

This made me double-check and it turns out it's because my hostname had an umlaut "o" (ö) in it. I changed my hostname and the problem has gone. Should this issue be closed then?

**Comment by mcornella** [comment]
Can you check if this fixes it? (just copy and paste to the terminal)
```zsh
  function update_terminalapp_cwd() {
    emulate -L zsh

    # Percent-encode the pathname.
    local URL="$(omz_urlencode -P "file://{$HOST}${PWD}")"
    [[ $? != 0 ]] && return 1

    # Undocumented Terminal.app-specific control sequence
    printf '\e]7;%s\a' "$URL"
  }
```

**Comment by mcornella** [comment]
Closing since this was solved in a2cad16790aeb7e775da47fd11f1f7cee1e03c90.
